File: Design/1396.design_underground_system.py
class UndergroundSystem:

    def __init__(self):
        self.customers = {}
        self.stations = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.customers[id] = [stationName, t]

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        startStation = self.customers[id][0]
        startTime = self.customers[id][1]
        self.customers[id] = []
        if startStation not in self.stations:
            self.stations[startStation] = {}
        if stationName not in self.stations[startStation]:
            self.stations[startStation][stationName] = [0, 0]
        prev_avg, prev_num = self.stations[startStation][stationName]
        update_num = prev_num + 1
        update_avg = (prev_avg * prev_num + t - startTime) / update_num
        self.stations[startStation][stationName] = [update_avg, update_num]

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        if startStation not in self.stations:
            self.stations[startStation] = {}
        if endStation not in self.stations[startStation]:
            self.stations[startStation][endStation] = [0, 0]
        return self.stations[startStation][endStation][0]


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: C++/Hash Table/1396. Design Underground System.cpp
class UndergroundSystem {
    
    unordered_map<int, pair<string, int>> customers;
    unordered_map<string, pair<int, int>> trips;
    
public:
    UndergroundSystem() {
        
    }
    
    void checkIn(int id, string stationName, int t) {
        customers[id].first = stationName;
        customers[id].second = t;
    }
    
    void checkOut(int id, string stationName, int t) {
        string trip = customers[id].first + "_" + stationName;
        int duration = t - customers[id].second;
        trips[trip].first++;
        trips[trip].second += duration;
    }
    
    double getAverageTime(string startStation, string endStation) {
        string trip = startStation + "_" + endStation;
        return 1.0 * trips[trip].second / trips[trip].first;
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */